networkx/algorithms/tests/test_similarity.py

```diff
@@ -43,6 +43,22 @@ def setup_class(cls):
         numpy = pytest.importorskip('numpy')
         scipy = pytest.importorskip('scipy')
 
+    def test_graph_edit_distance_roots_and_timeout(self):
+        G0 = nx.star_graph(5)
+        G1 = G0.copy()
+        pytest.raises(ValueError, graph_edit_distance, G0, G1, roots=[2])
+        pytest.raises(ValueError, graph_edit_distance, G0, G1, roots=[2, 3, 4])
+        pytest.raises(nx.NodeNotFound, graph_edit_distance, G0, G1, roots=(9, 3))
+        pytest.raises(nx.NodeNotFound, graph_edit_distance, G0, G1, roots=(3, 9))
+        pytest.raises(nx.NodeNotFound, graph_edit_distance, G0, G1, roots=(9, 9))
+        assert graph_edit_distance(G0, G1, roots=(1, 2)) == 0
+        assert graph_edit_distance(G0, G1, roots=(0, 1)) == 8
+        assert graph_edit_distance(G0, G1, roots=(1, 2), timeout=5) == 0
+        assert graph_edit_distance(G0, G1, roots=(0, 1), timeout=5) == 8
+        assert graph_edit_distance(G0, G1, roots=(0, 1), timeout=0.0001) is None
+        # test raise on 0 timeout
+        pytest.raises(nx.NetworkXError, graph_edit_distance, G0, G1, timeout=0)
```

Comment:
Just a question: does anyone know of a good way to test things like timeouts? I don't know of any pattern for reliably *testing* things that depend on dynamic metrics like execution time.

Reply:
not sure if this is what you're asking.. but on my end I was checking that the actual execution time does not exceed the timeout value.

Reply:
Right, but in order for the test to validate that the timeout keyword is having the intended effect for long-running `foo`, it is necessary that `foo(timeout=None)` will always take longer than `timeout`. This can get tricky as execution times depend on all sorts of things outside of the context of the test itself (platform, hardware, etc.) I'm not sure how to design a good test that is insensitive to all of these factors and reliably tests the timeout itself. Anyways - not a blocker here, but just curious if any has any good ideas!

Reply:
I thought that by setting a very very short timeout the first time the code runs the "prune" function that time will already have passed and no progress would have been made so there isn't any solution to return and ```None``` gets returned.  Probably not the best way, but it at least tests that the output is different from the output with a longer timeout.  I picked timeout=5 for the other cases because that should be long enough to get the full answer -- so that just tests that the timeout input is a valid input.

Did any of the travis-ci runs not pass these tests yet?  :}

Reply:
> Did any of the travis-ci runs not pass these tests yet? :}

Nope, tests are all green (for me locally and on CI) - this was more a high-level question on my part, just wondering if anyone had any insight into best-practices for testing things that depend on execution times. The current tests seem totally fine to me, sorry for dragging us off into the weeds :)